README.md
[![Circle CI](https://circleci.com/gh/rackspace-cookbooks/rackspace_apache_php.svg?style=svg)](https://circleci.com/gh/rackspace-cookbooks/rackspace_apache_php)

# rackspace_apache_php-cookbook

A cookbook to provide a web server able to serve php pages with Apache and PHP fpm.
It relies on [apache2 cookbook](https://github.com/svanzoest-cookbooks/apache2/) and [php-fpm](https://github.com/yevgenko/cookbook-php-fpm). Those cookbooks are pinned on well known working minor version to prevent breaking changes.
In addition (even if this is not a requirement), the cookbook will install php packages through [PHP cookbook](https://github.com/opscode-cookbooks/php). Indeed most of the time you will need `php::default` in your role which will conflict with the `php-fpm` package if they are different.
You can disable the installation of php packages with `node['rackspace_apache_php']['php_packages_install']['enable']`.

## ***
## NOTE: Support for PHP 5.4 was dropped in v1.0.0
## ***

## Supported Platforms

* Centos 6.7
* Ubuntu 12.04
* Ubuntu 14.04

## Supported PHP versions

* 5.5
* 5.6

## Attributes

* `node['rackspace_apache_php']['php_version']` : Which PHP version to install, default to PHP 5.6
* `node['rackspace_apache_php']['php-fpm']['default_pool']['enable']` : Should it enable a default PHP-FPM pool which listens on a unix socket, defaults to 'true' (change it to false if you want to manage your own PHP-FPM pools)
* `node['rackspace_apache_php']['php_handler']['enable']` : Should it enable Apache PHP handler (applied in "conf.d", so it will be available in EVERY vhost, if you want to manage your own handler configuration, set this attribute to false)
* `node['rackspace_apache_php']['php_handler']['cookbook']` : Where to find the handler configuration , default to `rackspace_apache_php cookbook`
* `node['rackspace_apache_php']['php_handler']['template']` : Where to find the handler configuration , default to `php-handler.conf.erb`

## Usage

Place a dependency on the rackspace_apache_php cookbook in your cookbook's metadata.rb
```
depends 'rackspace_apache_php'
```
Then, add `rackspace_apache_php::default` to your runlist

```
# myrecipe.rb
include_recipe 'rackspace_apache_php::default'
```

or

```
# roles/myrole.rb
name "myrole"
description "apache and php role"
run_list(
  "rackspace_apache_php::default"
)
```

You can change any of the `apache2`,`php-fpm` and `php` cookbook attributes to tune rackspace_apache_php configuration.
** However you should not change ** `['php-fpm']['package_name']`,`['php-fpm']['service_name']` or `['php']['packages']` (as they are part of this cookbook logic) without checking it works.

## In scope

The goal of this library is to do the basic configuration to serve PHP pages through Apache. It will only configure `apache2` and the default php handler, users are free to configure their vhost if they need anything more specific.

More in details it :

* Installs and configure Apache2 web server
* Installs and configure php-fpm
* Installs and configure php
* Configures Apache2 to serve php pages through php-fpm (in conf.d)
* Gets the correct packages and change the configuration according to the php/apache version

## Out of scope

Virtual Host are not managed by this cookbook, the configuration provided by this cookbook should not prevent users to extend php or php-fpm configuration.
As many features as possible should have a flag to enable/disable them, it will allow to enjoy some parts of the work done by this cookbook (get the correct packages by example) but still be able to configure your own php-fpm pools.


### Examples
#### Apache and PHP 5.5

```
node.default['rackspace_apache_php']['php_version'] = '5.5'
include_recipe 'rackspace_apache_php::default'
```

#### Apache and PHP 5.6

```
include_recipe 'rackspace_apache_php::default'
```

#### Apache and PHP 5.6 without default PHP handler and default PHP-FPM pool

You will have to add your own Vhost to configure the handler, here is an example using a `web_app` resource and the `apache_conf` definition from the upstream apache2 cookbook to create a virtual host and the Apache configuration necessary to handle php requests (templates not provided). It also leverages the `['php-fpm']['pools']` attribute from the upstream php-fpm cookbook to create a custom defined php-fpm pool listening on a socket. Although not provided below, the templates for the `apache_conf` definition and the `web_app` resource should contain the necessary directives to pass requestes for php files to the php-fpm pool.

```
node.default['rackspace_apache_php']['php_handler']['enable'] = false
node.default['rackspace_apache_php']['php-fpm']['default_pool']['enable'] = false

# Create a php-fpm pool through the attribute exposed from upstream
node.default['php-fpm']['pools'] = {
  override: {
    enable: 'true',
    process_manager: 'dynamic',
    max_requests: 5000
  }
}

include_recipe 'rackspace_apache_php::default'

# Create your own php-handler
apache_conf 'my-php-handler' do
  source 'my-php-handler.conf.erb'
  cookbook 'my-cookbook'
  enable true
end

# Create a virtual host to serve pages
web_app "my_site" do
  server_name node['hostname']
  server_aliases [node['fqdn'], "my-site.example.com"]
  docroot "/srv/www/my_site"
  cookbook 'my-cookbook'
  php_socket '/var/run/php-fpm-override.sock'
end

```

## References

* [Apache2 cookbook](https://github.com/svanzoest-cookbooks/apache2)
* [PHP-fpm cookbook](https://github.com/yevgenko/cookbook-php-fpm)
* [PHP cookbook](https://github.com/opscode-cookbooks/php)


## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: Julien Berard (julien.berard@rackspace.co.uk), Kostas Georgakopoulos (kostas.georgakopoulos@rackspace.co.uk)


CHANGELOG.md
rackspace_apache_php CHANGELOG
==================

1.0.0
-----
- Release as v1.0.0
- Dropped support for PHP 5.4
- Updated unit tests

0.2.2
-----
- Unit test fixes
- Support for CentOS 6.7 instead of 6.6

0.2.1
-----
- Fix for Rubocop tests
- Improvements on integration tests

0.2.0
-----
- Added support for mod_proxy_fgi when using Apache 2.4

0.1.1
-----
- Corrections and clean up on tests

0.1.0
-----
- Added feature to enable/disable a default php-fpm pool
- Disabled the default pool created by upstream php-fpm cookbook
- Added testing for an 'override' recipe
- Added 'supports' statements in metadata
- Various improvements in unit and integration tests
- Changed supported Centos from 6.5 to 6.6 in README

0.0.4
-----
- Pinned upstream apache2, php and php-fpm cookbooks to minor versions with '='
- Enabled the default vhost from upstream apache2 only for integration tests

0.0.3
-----
- Removed unneeded 'libraries' directory

0.0.2
-----
- Added support for PHP in addition of PHP-fpm

0.0.1
-----
- First version


recipes/default.rb
# Encoding: utf-8
#
# Cookbook Name:: rackspace_apache_php
# Recipe:: default
#
# Copyright 2014, Rackspace
#

# We only support 5.5 and 5.6
if node['rackspace_apache_php']['php_version'] != '5.5' && node['rackspace_apache_php']['php_version'] != '5.6'
  Chef::Application.fatal!("PHP version #{node['rackspace_apache_php']['php_version']} is not supported")
end

include_recipe 'chef-sugar'

# APACHE
include_recipe 'apache2::mod_actions'
include_recipe 'apache2::mod_fastcgi' unless ubuntu_trusty? # On Ubuntu 14.04 we are going to use mod_proxy_fcgi which is bundled with Apache 2.4

if ubuntu_trusty? # The required modules for Trusty are installed along with Apache 2.4 but not enabled by default
  %w('proxy' 'proxy_fcgi').each do |mod|
    apache_module mod
  end
end

apache_conf 'php-handler' do
  source node['rackspace_apache_php']['php_handler']['template']
  cookbook node['rackspace_apache_php']['php_handler']['cookbook']
  enable node['rackspace_apache_php']['php_handler']['enable']
end

# PHP-FPM
# repo dependencies for php-fpm
#
php_fpm = {
  'rhel' => {
    '5.5' => {
      'php_packages' => ['php55u-pear', 'php55u-devel', 'php55u-cli', 'php55u-pear'],
      'php_fpm_package' => 'php55u-fpm',
      'service' => 'php-fpm'
    },
    '5.6' => {
      'php_packages' => ['php56u-pear', 'php56u-devel', 'php56u-cli', 'php56u-pear'],
      'php_fpm_package' => 'php56u-fpm',
      'service' => 'php-fpm'
    }
  },
  'debian' => {
    '5.5' => {
      'repo'    => 'http://ppa.launchpad.net/ondrej/php5/ubuntu',
      'php_packages' => ['php5-cgi', 'php5', 'php5-dev', 'php5-cli', 'php-pear'],
      'php_fpm_package' => 'php5-fpm',
      'service' => 'php5-fpm'
    },
    '5.6' => {
      'repo'    => 'http://ppa.launchpad.net/ondrej/php5-5.6/ubuntu',
      'php_packages' => ['php5-cgi', 'php5', 'php5-dev', 'php5-cli', 'php-pear'],
      'php_fpm_package' => 'php5-fpm',
      'service' => 'php5-fpm'
    }
  }
}

# Platform PHP repositories
# RHEL
if platform_family?('rhel')
  include_recipe 'yum-ius'
elsif platform_family?('debian')
  # DEBIAN
  include_recipe 'apt'
  # using http://ppa rather than ppa: to be sure it passes firewall
  apt_repository "php-#{node['rackspace_apache_php']['php_version']}" do
    uri          php_fpm[node['platform_family']][node['rackspace_apache_php']['php_version']]['repo']
    keyserver    'hkp://keyserver.ubuntu.com:80'
    key          '14AA40EC0831756756D7F66C4F4EA0AAE5267A6C'
    components   ['main']
    distribution node['lsb']['codename']
  end
  # however we don't want apache from ondrej/php5
  apt_preference 'apache' do
    glob         '*apache*'
    pin          'release o=Ubuntu'
    pin_priority '600'
  end
end

# PHP-FPM
# Set the correct php-fpm packages to install
node.default['php-fpm']['package_name'] = php_fpm[node['platform_family']][node['rackspace_apache_php']['php_version']]['php_fpm_package']
node.default['php-fpm']['service_name'] = php_fpm[node['platform_family']][node['rackspace_apache_php']['php_version']]['service']
include_recipe 'php-fpm::default'

# Create (or not) our default pool
# On Trusty we use mod_proxy_fcgi and we need Apache > 2.4.9 to use sockets (only available from ondrej which we don't use for Apache)
listen_on = ubuntu_trusty? ? '127.0.0.1:9000' : '/var/run/php-fpm-default.sock'
php_fpm_pool 'default' do
  listen listen_on
  enable node['rackspace_apache_php']['php-fpm']['default_pool']['enable']
end

# PHP
# Set the correct php packages to install
if node['rackspace_apache_php']['php_packages_install']['enable']
  node.default['php']['packages'] = php_fpm[node['platform_family']][node['rackspace_apache_php']['php_version']]['php_packages']
  include_recipe 'php::default'
  # Emptying php packages, to be sure if another cookbook tries to include php::default it
  # will not install unwanted php version. If a user decided to use override we assume he knows what he is doing.
  node.default['php']['packages'] = []
end

# TODO
# * downgrade / upgrade Apache if required
# Refactor mapping outside of the recipe
